### PR TITLE
Initial support for SAM3X8E on Arduino Due

### PR DESCRIPTION
**Pull request by BrentSeidel**
Changes and additions to provide support for ravenscar and zfp profiles on the Arduino Due.  Initial changes based off of the SAM4S profile.

This works well enough that I am able to load code and get the built-in LED to flash.  I wouldn't be surprised if there were some issues to fix, but it is a starting point.

This is my first submission, so please let me know if I need to do anything different.

thanks,
brent

### CODE REVIEW

**Comment by Fabien-Chouteau** [comment]
Thank you for the contribution @BrentSeidel !

I will have a look.

File: arm/cortexm.py [hunk 1]
```diff
@@ -154,6 +154,7 @@ def __init__(self):
 
         self.add_sources('crt0', [
             'arm/sam/s-sam4s.ads',
+            'arm/sam/s-sam3x8.ads',
```
Comment (Fabien-Chouteau) [inline arm/cortexm.py]: There should be a condition here. `arm/sam/s-sam3x8.ads` is only for the `s-sam3x8.ads` MCU.

**Comment by BrentSeidel** [comment]
Hi Fabian,

For your last item, adding a condition for the sam3x8e, I tried changing it to:
        if self.board == 'sam3x8s':
            self.add_sources('crt0', [
                'arm/sam/s-sam4s.ads',
                'arm/sam/s-sam3x8.ads',
                'arm/sam/start-rom.S',
                'arm/sam/start-ram.S',
                'arm/sam/setup_pll.ads'])
        else:
            self.add_sources('crt0', [
                'arm/sam/s-sam4s.ads',
                'arm/sam/start-rom.S',
                'arm/sam/start-ram.S',
                'arm/sam/setup_pll.ads'])

but when I try and run it, I get told that 'SamCommonArchSupport' object has no attribute ‘board’, is there another attribute that I can use for this?

thanks,
brent

**Comment by Fabien-Chouteau** [comment]
> but when I try and run it, I get told that 'SamCommonArchSupport' object has no attribute ‘board’, is there another attribute that I can use for this?

In that case you can move the condition to the Sam class. and just for the `s-samX.ads` file, not the others:
```python
if self.board == 'sam3x8s':
    self.add_sources('crt0', ['arm/sam/s-sam3x8s.ads'])
else:
    self.add_sources('crt0', ['arm/sam/s-sam4s.ads'])
```

**Comment by BrentSeidel** [comment]
I think that I have all the changes made and submitted.

**Comment by BrentSeidel** [comment]
I think that I've addressed all your listed issues.  Is there anything else that I need to do?

File: arm/cortexm.py [hunk 2]
```diff
@@ -194,13 +195,18 @@ def compiler_switches(self):
...
-        assert board in ('sam4s', 'samg55'), "Unexpected SAM board %s" % board
+        assert board in ('sam3x8e', 'sam4s', 'samg55'), "Unexpected SAM board %s" % board
         self.board = board
+        print self.board
         super(Sam, self).__init__()
 
         self.add_linker_script(
             'arm/sam/%s/memory-map.ld' % self.name,
             loader=('SAMBA', 'ROM'))
+        if board == 'sam3x8s':
```
Comment (Fabien-Chouteau) [inline arm/cortexm.py]: There is a typo here.

**Comment by Fabien-Chouteau** [comment]
Hi @BrentSeidel,

Did you follow the conversation on #12? 

What is the number of interrupt level on the sam3x8, and did you try to do some interrupt handling?

**Comment by BrentSeidel** [comment]
Hi @Fabien-Chouteau,

I corrected the typos.

As far as interrupts go, the data sheet says that the sam3A/X series have up to 30 interrupts with priorities 0-15.  This should be consistent with the default assumption.

I have been working on an interrupt driven serial port driver that works fairly well to transmission (it's in repository Ada-Arduino-Due).  I'm trying to add receive to it and it acts a little flakey, so there are some things that need to be fixed to make it more robust.  I've also had to learn about suspension_objects...

brent

**Comment by BrentSeidel** [comment]
By the way, how much of the standard Ada library is supported by this?  I can understand stuff like unbounded_string not being supported, but I just tried Integer'Value() and got told, "construct not allowed in configurable run-time mode".  I can roll my own integer conversion, if I have to, but I'm lazy.

For the Arduino Due with 512K flash and 96K RAM, one might be able to configure more of the library to be available.

**Comment by Fabien-Chouteau** [comment]
Hi @BrentSeidel,

> By the way, how much of the standard Ada library is supported by this?

You can have a look at the list of packages in the run-time install directory.
 
 > I can understand stuff like unbounded_string not being supported, but I just tried Integer'Value() and got told, "construct not allowed in configurable run-time mode". I can roll my own integer conversion, if I have to, but I'm lazy.
> 

Integer'Value should be supported on all run-times so I think there is a problem here.

> For the Arduino Due with 512K flash and 96K RAM, one might be able to configure more of the library to be available.

There are 3 run-time profiles (ZFP, Ravenscar-SFP and Ravenscar-Full). The features supported by each profile is defined by AdaCore and cannot be changed with the tools in this repository.

**Comment by BrentSeidel** [comment]
> Hi @BrentSeidel,
> 
> > By the way, how much of the standard Ada library is supported by this?
> 
> You can have a look at the list of packages in the run-time install directory.
> 
> > I can understand stuff like unbounded_string not being supported, but I just tried Integer'Value() and got told, "construct not allowed in configurable run-time mode". I can roll my own integer conversion, if I have to, but I'm lazy.
> 
> Integer'Value should be supported on all run-times so I think there is a problem here.
> 
> > For the Arduino Due with 512K flash and 96K RAM, one might be able to configure more of the library to be available.
> 
> There are 3 run-time profiles (ZFP, Ravenscar-SFP and Ravenscar-Full). The features supported by each profile is defined by AdaCore and cannot be changed with the tools in this repository.

Integer'Value() seems to require s-valint.ads.  This seems to be added in the sources.py file based on the 'Add_Value_Int' profile/scenario.  I haven't yet found out where this gets defined, but I'm still poking around.

In cortex.py, in the Sam(ArmV7MTarget) class, the system_ads property doesn't include Ravenscar-Full and has a comment, "# No runtime full".  It's not clear why it's not present.

**Comment by BrentSeidel** [comment]
I've encountered a problem.  I tried building the runtime with the latest submitted changes.  It built and installed without any errors/warnings.  However, when I try to build my application program (code is on https://github.com/BrentSeidel/Ada-Arduino-Due), I get a bunch of link errors to the effect that some object module in the runtime uses "VFP register arguments" while my main program does not.  Googling suggests that these messages indicate some inconsistencies in FPU use.  I've tried a variety of things to enable or disable floating point, but nothing seems to work.  This seems odd to me as I would expect that the runtime would tell the compiler what FPU options (none, software, hardware) to use so that my application would be consistent with the runtime.  So, I am stumped.  Any suggestions?

On the plus side, I think that I have a Sam3x8e version of s-textio, but I can't test it until I get the FPU problem resolved.

thanks,
brent

**Comment by BrentSeidel** [comment]
I have found a solution to get Integer'Value() to work on the ravenscar-sfp profile.  In bb-runtimes/support/rts_sources/profiles.py, somewhere around line 96 in the sfp_scenarios function definition, add the following lines:
        ret['Add_Value_Int'] = "yes"
        ret['Add_Value_LL_Int'] = "yes"
        ret['Add_Value_Based_Int'] = "yes"
        ret['Add_Value_LL_Based_Int'] = "yes"
The first one may be the only one that's really needed.  One could do some experimenting to see what other features could be added here.

brent

**Comment by BrentSeidel** [comment]
Another question:  Is 64 bit integer arithmatic suppose to be part of all runtimes?  My BME280 driver uses it.  It could probably be worked around, but would be nice not to have to.

thanks
brent

**Comment by Fabien-Chouteau** [comment]
> Is 64 bit integer arithmatic suppose to be part of all runtimes?

Yes it should work.

**Comment by BrentSeidel** [comment]
Hi @FabienChouteau,

I have a simple patch that enables Integer’Value() and 64 bit arithmetic for the Ravenscar SFP.. In bb-runtimes/support/rts_sources/profiles.py, somewhere around line 96 in the sfp_scenarios function definition, add the following lines:
ret['Add_Value_Int'] = "yes"
ret['Add_Value_LL_Int'] = "yes"
ret['Add_Value_Based_Int'] = "yes"
ret['Add_Value_LL_Based_Int'] = “yes"
ret['Add_Arith64'] = “yes”

There may be a better way to do this by playing around with the conditions in sources.py, but I haven’t tried touching that yet.  I’ve been busy actually programming my Arduino Due.  I’ve got the serial ports working with both polled and quite a nice interrupt driven driver that support character echo and backspace/delete.  The I2C ports are working.  Analog inputs are working, and I’m working on analog outputs right now.  The code for these is on GitHub at BrentSeidel/Ada-Arduino-Due.

brent

**Comment by Electromechanical-Supergiant** [comment]
What changes need to happen still to get this merged? I am also interested in having the Arduino Due supported and I'd like to help get this PR to a state where it can be merged.

Also, this has not been addressed yet:

>In cortex.py, in the Sam(ArmV7MTarget) class, the system_ads property doesn't include Ravenscar-Full and has a comment, "# No runtime full". It's not clear why it's not present.

I'd be interested in using the full Ravenscar runtime, so I'd also like to know what can be done to make that happen.

**Comment by Fabien-Chouteau** [comment]
Hi @Rodeo-McCabe,

I need to find some time to review this again.

**Comment by BrentSeidel** [comment]
Hi @Rodeo-McCabe ,

Thanks for your interest in my project.  I would also like to get this finalized.  However I have started a new contract which leaves me little time to push this and in most of my available time I've been working on device drivers on the Arduino Due (see my repositories BBS-BBB-Ada and Ada-Arduino-Due).

**Comment by Fabien-Chouteau** [comment]
Hello @BrentSeidel, can you rebase your changes on the community-2020 branch?

**Comment by BrentSeidel** [comment]
@Fabien-Chouteau I think that I've managed to rebase it.  But it's not building yet.  It's been over a year since I've looked at this so it'll probably take me a few days (at least) to get things figured out again.

**Comment by Fabien-Chouteau** [comment]
Alright, let us know and open a new PR.

Thanks,

**Pull request closed without merging**